lib/puppet/face/config.rb

```diff
@@ -106,6 +106,17 @@
     EOT
 
     when_invoked do |name, value, options|
+      unless name == 'environment' && (options[:section] == 'agent' || options[:section] == 'user')
+        Puppet.warning _(<<-EOM).chomp % {}
+The environment should be set in either the [user] or [agent] section. Variables
+set in the agent section are used when running `puppet agent`. Variables set in
+the user section are used when running various other puppet subcommands, like
+`puppet apply` and `puppet module`. Do this by adding `--section user` or
+`--section agent` to the `puppet config` command. For more information, see
```

Comment:
I found the sentence "Do this by adding ..." confusing when I first read it.

Maybe something like:

To print or set the setting use the `puppet config` command by adding the `--section user` or `--section agent` options.

Reply:
I agree it's not the easiest sentence to parse.

However, users will only get this error message when using `puppet config set`. How about the following rewrite:

Set the config section by using the `--section` flag. For example, `puppet config --section user set environment foo`

Reply:
👍

Reply:
Updated.